chapter-1/src/parse.c

```diff
@@ -189,181 +144,117 @@ TokenFinder makeNumberFinder() {
+	.token = T_IDENT,
+	.transitionCount = ARRAY_LENGTH(transitions),
+	.transitions = transitions,
     };
-
-    return charFinder;
 }
 
-TokenFinder makeIdentifierFinder() {
-    CharState charState;
-    charState.id = 0;
-    charState.type = LETTER;
-
-    CharState hyphenState;
-    hyphenState.id = 1;
-    hyphenState.type = HYPHEN;
+static const StateTransition openParenTransitions[] = {
+    {
+	.fromState = startState,
+	.toState = { .id = 0, .type = OPEN_PAREN }
+    },
+    {
+	.fromState = { .id = 0, .type = OPEN_PAREN },
+	.toState = endState
+    }
+};
+
+static const TokenFinder openParenFinder = {
+    .token = T_OPEN_PAREN,
+    .transitionCount = ARRAY_LENGTH(openParenTransitions),
+    .transitions = openParenTransitions
+};
+
+static const StateTransition closeParenTransitions[] = {
+    {
+	.fromState = startState,
+	.toState = { .id = 0, .type = CLOSE_PAREN }
+    },
+    {
+	.fromState = { .id = 0, .type = CLOSE_PAREN },
+	.toState = endState
+    }
+};
+
+static const TokenFinder closeParenFinder = {
+    .token = T_CLOSE_PAREN,
+    .transitionCount = ARRAY_LENGTH(closeParenTransitions),
+    .transitions = closeParenTransitions
+};
+
+static const StateTransition whitespaceTransitions[] = {
+    {
+	.fromState = startState,
+	.toState = { .id = 0, .type = SPACE }
+    },
+    {
+	.fromState = { .id = 0, .type = SPACE },
+	.toState   = { .id = 0, .type = SPACE }
+    },
+    {
+	.fromState = { .id = 0, .type = SPACE },
+	.toState = endState
+    }
+};
 
-    TokenFinder identFinder;
-    identFinder.token = T_IDENT;
-    identFinder.transitionCount = 7;
-    identFinder.transitions = malloc(sizeof(StateTransition) * identFinder.transitionCount);
+static const TokenFinder whitespaceFinder = {
+    .token = T_WHITESPACE,
+    .transitionCount = ARRAY_LENGTH(whitespaceTransitions),
+    .transitions = whitespaceTransitions
+};
 
-    identFinder.transitions[0] = (StateTransition) {
-        .fromState = startState,
-        .toState = charState
-    };
-    identFinder.transitions[1] = (StateTransition) {
-        .fromState = startState,
-        .toState = hyphenState
-    };
-    identFinder.transitions[2] = (StateTransition) {
-        .fromState = charState,
-        .toState = charState
-    };
-    identFinder.transitions[3] = (StateTransition) {
-        .fromState = charState,
-        .toState = hyphenState 
-    };
-    identFinder.transitions[4] = (StateTransition) {
-        .fromState = charState,
-        .toState = endState
-    };
-    identFinder.transitions[5] = (StateTransition) {
-        .fromState = hyphenState,
-        .toState = charState
-    };
-    identFinder.transitions[6] = (StateTransition) {
-        .fromState = hyphenState,
-        .toState = endState
+void initTokenFinders() {
+    TokenFinder finders[] = {
+      makeNumberFinder(),
+      whitespaceFinder,
+      openParenFinder,
+      closeParenFinder,
+      makeIdentifierFinder(),
+      makeCellrefFinder(),
```

Comment:
We can actually get rid of this one altogether. It's left over from the previous project I copied the tokenizer from.

Reply:
Cool - removed.